---
title: "GCclassifier: An R package for the prediction of molecular subtypes of gastric cancer"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: TRUE
bibliography: bibliography.bib
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{GCclassifier: An R package for the prediction of molecular subtypes of gastric cancer}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  fig.height = 9,
  fig.align = 'center',
  collapse = TRUE,
  message = TRUE,
  echo = TRUE,
  warnings = FALSE
)
```

Jiang Li^1,2^, Xin Wang^1,3^

^1^ Department of Surgery, The Chinese University of Hong Kong, Hong Kong SAR, China.\
^2^ Department of Biomedical Sciences, City University of Hong Kong, Hong Kong SAR, China.\
^3^ Shenzhen Research Institute, The Chinese University of Hong Kong, Shenzhen, China.

-   Contact: [xinwang\@cuhk.edu.hk](mailto:xinwang@cuhk.edu.hk){.email}   
-   Package: `r BiocStyle::pkg_ver("GCclassifier")`

# Introduction

Gastric cancer (GC) is one of the most commonly diagnosed malignancies, threatening millions of lives worldwide each year. Importantly, GC is a heterogeneous disease, posing a significant challenge to the selection of patients for more optimized therapy. Over the last decades, extensive community effort has been spent on dissecting the heterogeneity of GC, leading to the identification of distinct molecular subtypes that are clinically relevant. However, so far, no tool is publicly available for GC subtype prediction, hindering the research into GC subtype-specific biological mechanisms, the design of novel targeted agents, and potential clinical applications. To address the unmet need, we developed an R package `GCclassifier` for predicting GC molecular subtypes based on gene expression profiles. The predictive performance of `GCclassifier` was demonstrated using case studies on multiple independent datasets.

The `GCclassifier` package enables the molecular subtype prediction of GC through log2 scaled gene expression profiles, employing the random forest algorithm (https://www.stat.berkeley.edu/~breiman/RandomForests/). Additionally, it offers non-R users the convenience of utilizing local and online Shiny applications for streamlined GC molecular subtype prediction. The package installation requires a minimum R version of 4.1.0.

# Input data

GCclassifier demonstrates robust prediction performance on both microarray and RNA-seq datasets. It is imperative that gene expression profiles employ gene IDs from authoritative sources such as NCBI Entrez, Ensembl, HGNC symbol, or Refseq. Furthermore, the colnames should be appropriately designated as sample names to ensure accurate processing and analysis. More specifically:

-   Microarray data input should be pre-processed and normalized (log2 transformed).\
-   RNA-sequencing data input should be log2 transformed normalized values (log2 TPM recommended).\
-   At least two samples were required in the microarray/RNA-seq data.

When the input dataset lacks log2 transformation, the function will automatically detect the range of gene expression profiles and execute the log2 transformation, if necessary.

# Quick start

## Pre-requisite of installation

Prior to executing the example codes in the vignette, the following R packages are necessary:

-   dplyr (dataframe manipulation)
-   magrittr (pipe operator)
-   randomForest (prediction classifier)
-   BiocManager (package installation)
-   shiny (Shiny webpage)
-   DT (interactive tables in Shiny)
-   shinyjs (shiny webpage)
-   impute (missing data imputation)
-   AnnotationDbi (gene identifier converter)
-   org.Hs.eg.db (gene identifier converter)

```{r, eval=FALSE}
# Required packages: run if not already installed
if(!requireNamespace('BiocManager')){
  install.packages('BiocManager')
}

BiocManager::install(c(
  'impute', 'dplyr', 'magrittr', 'AnnotationDbi', 'randomForest', 'org.Hs.eg.db', 'shiny', 'DT', 'shinyjs'
)ï¼Œ force = T)
```

## Load example dataset GSE62254

In this case study, we employed the `GCclassifier` package to conduct GC molecular subtype prediction based on gene expression profiles. The gene expression profiles used in this analysis belong to the ACRG cohort, derived from a microarray dataset comprising 300 GC patients known as [GSE62254](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE62254) (@Cristescu2015). Subsequently, we utilized these gene expression profiles to predict the molecular subtype using three distinct subtype systems, namely `EMP`(@Oh2018), `ACRG`(@Cristescu2015), and `TCGA`(@TCGA2014).

```{r, eval=TRUE}
  ## Load example data set from GCclassifier and have a glimpse
  library(GCclassifier)
  data("GSE62254")
  data("GSE62254_subtype")
  GSE62254[1:5, 1:5]
  
  ## Check whether sample names in expression profile is identical to the sample names in clinical information
  identical(colnames(GSE62254), GSE62254_subtype$GEO_ID)
```

The gene expression profiles were acquired using the R package `GEOquery`, and subsequently, the expression levels were mapped from probes to gene symbols. In cases where multiple probes were associated with the same gene, the probe featuring the maximum expression value was selected for further analysis. In the GSE62254 dataset, the `rownames(GSE62254)` represented gene symbols, while the `colnames(GSE62254)` denoted the sample names.

## Case study 1: The EMP subtype prediction

To predict the EMP subtypes, the gene expression profiles from GSE62254 serve as the input data, while additional information regarding other parameters can be found in the help page of the `classifyGC` function:

```{r, eval=TRUE, message=TRUE}
  emp.res <- classifyGC(
    Expr = GSE62254, ## gene expression profile with log2 transformation
    method = 'EMP', ## subtyping system
    idType = 'SYMBOL', ## the gene identifier type in gene expression profile
    useMinPosterior = F ## whether the minPosterior threshold is also used for EMT subtyping
  )
  
  table(emp.res$subtype, GSE62254_subtype$Subgroup)
```

It is important to note that the useMinPosterior parameter plays a crucial role in stratifying subtypes for the EMT subtype. By default, the MP and EP subtype labels are determined using the recommended pre-trained Youden index value. Nonetheless, users have the flexibility to modify the minPosterior and useMinPosterior parameters to optimize their specific results. However, when applying the TCGA subtyping system, minPosterior functions differently, as it is utilized to classify samples as Unclassified only if the probabilities of all four subtypes fall below this specified value.

Subsequent to predicting the EMP subtypes of GSE62254, the survival outcomes of the predicted labels would be employed to assess the clinical relevance of the subtypes. For this purpose, two essential packages, namely `survival` and `survminer`, need to be installed prior to the analysis:

```{r, eval=TRUE, message=FALSE}
  if(!requireNamespace('survival')){
    BiocManager::install('survival')
  }

  if(!requireNamespace('survminer')){
    BiocManager::install('survminer')
  }

  library(survival)
  library(survminer)
  GSE62254_subtype[['EMP']] <- emp.res$subtype
  ggsurvplot(
    survfit(Surv(OS.m, Death) ~ EMP, data = GSE62254_subtype),
    pval = TRUE,
    conf.int = F,
    risk.table = T,
    risk.table.col = "strata",
    size = 1.5 ,
    pval.size = 8,
    xlab = 'Follow up (months)',
    ylab = 'Overall survival (%)',
    legend.title = '',
    legend.lab = c('EP', 'MP'),
    ggtheme = theme_classic2(),
    palette = "Dark1"
  ) + ggtitle('GSE62254 - OS')
  
  ggsurvplot(
    survfit(Surv(DFS.m, Recur) ~ EMP, data = GSE62254_subtype),
    pval = TRUE,
    conf.int = F,
    risk.table = T,
    risk.table.col = "strata",
    size = 1.5,
    pval.size = 8,
    xlab = 'Follow up (months)',
    ylab = 'Disease-free survival (%)',
    legend.title = '',
    legend.lab = c('EP', 'MP'),
    ggtheme = theme_classic2(),
    palette = "Dark1"
  ) + ggtitle('GSE62254 - DFS')
```

## Case study 2: The ACRG subtype prediction

To predict the ACRG subtypes, the procedure remains the same, with the only variation being the alteration of the `method` parameter from EMP to ACRG. The rest of the analysis remains unchanged.

```{r, eval=TRUE, message=TRUE}
  acrg.res <- classifyGC(
    Expr = GSE62254, 
    method = 'ACRG', 
    idType = 'SYMBOL'
  )
  table(acrg.res$subtype, GSE62254_subtype$ACRG.sub)
```

```{r, eval=TRUE, message=FALSE}
GSE62254_subtype$ACRG <- acrg.res$subtype
survfit(Surv(OS.m, Death) ~ ACRG, data = GSE62254_subtype) %>%
  ggsurvplot(
    pval = TRUE,
    conf.int = F,
    risk.table = T,
    risk.table.col = "strata",
    size = 1.5 ,
    pval.size = 8,
    xlab = 'Follow up (months)',
    ylab = 'Overall survival (%)',
    legend.title = '',
    legend.lab = c('MSI', 'MSS/EMT', 'MSS/TP53-', 'MSS/TP53+'),
    ggtheme = theme_classic2(),
    palette = "Dark1"
  ) + ggtitle('GSE62254')

survfit(Surv(DFS.m, Recur) ~ ACRG, data = GSE62254_subtype) %>%
  ggsurvplot(
    pval = TRUE,
    conf.int = F,
    risk.table = T,
    risk.table.col = "strata",
    size = 1.5,
    pval.size = 8,
    xlab = 'Follow up (months)',
    ylab = 'Disease-free survival (%)',
    legend.title = '',
    legend.lab = c('MSI', 'MSS/EMT', 'MSS/TP53-', 'MSS/TP53+'),
    ggtheme = theme_classic2(),
    palette = "Dark1"
  ) + ggtitle('GSE62254')
```

## Case study 3: The TCGA subtype prediction

To predict the TCGA subtypes, set the `method` parameter to `TCGA`, the rest parameters are the same:

```{r, eval=TRUE, message=TRUE}
  tcga.res <- classifyGC(
    Expr = GSE62254,
    method = 'TCGA',
    idType = 'SYMBOL'
  )
```

```{r, eval=TRUE, message=FALSE}
GSE62254_subtype$TCGA <- tcga.res$subtype
survfit(Surv(OS.m, Death) ~ TCGA, data = GSE62254_subtype) %>%
  ggsurvplot(
    pval = TRUE,
    conf.int = F,
    risk.table = T,
    risk.table.col = "strata",
    size = 1.5 ,
    pval.size = 8,
    xlab = 'Follow up (months)',
    ylab = 'Overall survival (%)',
    legend.title = '',
    legend.lab = c('CIN', 'EBV', 'GS', 'MSI'),
    ggtheme = theme_classic2(),
    palette = "Dark1"
  ) + ggtitle('GSE62254')

survfit(Surv(DFS.m, Recur) ~ TCGA, data = GSE62254_subtype) %>%
  ggsurvplot(
    pval = TRUE,
    conf.int = F,
    risk.table = T,
    risk.table.col = "strata",
    size = 1.5,
    pval.size = 8,
    xlab = 'Follow up (months)',
    ylab = 'Disease-free survival (%)',
    legend.title = '',
    legend.lab = c('CIN', 'EBV', 'GS', 'MSI'),
    ggtheme = theme_classic2(),
    palette = "Dark1"
  ) + ggtitle('GSE62254')
```

# The Shiny application
In order to streamline the process for non-R users and alleviate programming complexities, a Shiny application for molecular subtype prediction using gene expression profiles has been made available. By executing the `classifyGC_interface()` function, the Shiny application will automatically open on the default browser (Chrome, Firefox, Edge, and Safari recommended). Additionally, an example dataset is accessible on the `Analyze` page, and the generated subtype prediction results can be directly downloaded in various export formats. For further details, please run `??classifyGC_interface` to access comprehensive information.

```{r, eval=FALSE}
  # Execute this function to initiate the internal Shiny application
  classifyGC_interface()
```

It is important to note that the errors encountered while running the Shiny application may vary due to the intricate nature of the uploaded gene expression profiles. Considering the complexity of potential errors, attempting to address all scenarios is impractical. To mitigate issues, users are strongly advised to adhere to the rules specified in the example dataset and exercise caution in ensuring compliance with the requirements for uploading gene expression profiles:

-   Currently, the expression profiles with 'NA' values cannot be uploaded and processed by GCclassifier. It is suggested to perform imputation before uploading to the application.\
-   Please choose the right type of gene identifiers according to your data. Currently, GCclassifier can accept NCBI Entrez, Ensembl, HGNC symbol and Refseq.\
-   Keywords such as "SYMBOL", "ENSEMBL", "ENTREZID" and "REFSEQ" cannot be included in the file columns.   
-   Only numeric values in gene expression profiles are accepted.\
-   Gene expression profiles should not contain any negative value(s).   

In the event of encountering error messages while using the Shiny web application, it is crucial to meticulously review the uploaded gene expression profiles and strictly adhere to the format outlined in the example dataset. Paying close attention to these details will help in resolving any issues effectively.   

# Session Info
```{r, echo=FALSE}
  sessionInfo()
```

# References
